net/udprelay/server.go

```diff
@@ -0,0 +1,517 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package udprelay
+
+import (
+	"bytes"
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"net"
+	"net/netip"
+	"strconv"
+	"sync"
+	"time"
+
+	"go4.org/mem"
+	"tailscale.com/disco"
+	"tailscale.com/net/packet"
+	"tailscale.com/types/key"
+)
+
+const (
+	defaultBindLifetime        = time.Second * 5
+	defaultSteadyStateLifetime = time.Minute * 5
+)
+
+// Server implements an experimental UDP relay server.
+type Server struct {
+	// disco keypair used as part of 3-way bind handshake
+	disco       key.DiscoPrivate
+	discoPublic key.DiscoPublic
+
+	bindLifetime        time.Duration
+	steadyStateLifetime time.Duration
+
+	// addrPorts contains the ip:port pairs returned as candidate server
+	// endpoints in response to an allocation request.
+	addrPorts []netip.AddrPort
+
+	uc *net.UDPConn
+
+	closeOnce sync.Once
+	wg        sync.WaitGroup
+	closeCh   chan struct{}
+	closed    bool
+
+	mu        sync.Mutex // guards the following fields
+	lamportID uint64
+	vniPool   []uint32 // the pool of available VNIs
+	byVNI     map[uint32]*serverEndpoint
+	byDisco   map[pairOfDiscoPubKeys]*serverEndpoint
+}
+
+// pairOfDiscoPubKeys is a pair of key.DiscoPublic. It must be constructed via
+// newPairOfDiscoPubKeys to ensure lexicographical ordering.
+type pairOfDiscoPubKeys [2]key.DiscoPublic
+
+func (p pairOfDiscoPubKeys) String() string {
+	return fmt.Sprintf("%s <=> %s", p[0].ShortString(), p[1].ShortString())
+}
+
+func newPairOfDiscoPubKeys(discoA, discoB key.DiscoPublic) pairOfDiscoPubKeys {
+	var pair pairOfDiscoPubKeys
+	cmp := discoA.Compare(discoB)
+	if cmp == 1 {
+		pair[0] = discoB
+		pair[1] = discoA
+	} else {
+		pair[0] = discoA
+		pair[1] = discoB
+	}
+	return pair
+}
+
+// ServerEndpoint contains the Server's endpoint details.
+type ServerEndpoint struct {
+	// ServerDisco is the Server's Disco public key used as part of the 3-way
+	// bind handshake. Server will use the same ServerDisco for its lifetime.
+	// ServerDisco value in combination with LamportID value represents a
+	// unique ServerEndpoint allocation.
+	ServerDisco key.DiscoPublic
+
+	// LamportID is unique and monotonically increasing across ServerEndpoint
```

Comment:
nit: "monotonically non-decreasing" is probably the constraint we want (may be allowed to have gaps, but never decrease or be reused)

Reply:
is this intended to be ephemeral for server restarts? if so do we need to anchor it to another comparable value to differentiate restarts?

Reply:
Yes, take a look at the comment on the `ServerDisco` field.

Reply:
Updated the doc in https://github.com/tailscale/tailscale/pull/15480/commits/c2f7d7aad7dc5099b0e194ac491820271d6806ec.